### DISCUSSION TITLE
A problem about drag widget or window

### DISCUSSION THREAD

**Discussion by Horbin-Magician** [slint-ui/slint#3355] [Q&A] [answered]
I want to implement a draggable window. But when I drag the window, it starts to shake strangely.

Here is a simple demo to show the problem：

```Rust
fn main() {
    let pin_win = PinWindow::new().unwrap();
    
    pin_win.window().set_position( slint::LogicalPosition::new(0., 0.));
    pin_win.set_now_pos( slint::LogicalPosition::new(0., 0.) );

    let pin_win_clone = pin_win.as_weak();
    pin_win.on_mouse_move(move |delta_x, delta_y| {
        let pin_win_clone = pin_win_clone.unwrap();
        let now_pos = pin_win_clone.get_now_pos();
        pin_win_clone.window().set_position(
            slint::LogicalPosition::new(now_pos.x + delta_x, now_pos.y + delta_y)
        );
    });

    pin_win.run().unwrap();
}

slint::slint! {
    export component PinWindow inherits Window {
        no-frame: true;
        always-on-top: true;
        background: red;

        callback mouse_move(length, length);
        in-out property <Point> now_pos;
        in-out property <Point> mouse_down_pos;

        TouchArea {
            // property <bool> applying: false;  // dirty fix issue with double applying between event loops
            moved => {
                // if (self.applying) {                      // dirty fix
                    root.mouse_move(self.mouse-x - self.pressed-x, self.mouse-y - self.pressed-y);
                // } else {                                       // dirty fix
                //     self.applying = true;             // dirty fix
                // }                                                // dirty fix
            }
            pointer-event(event) => {
                if(event.button == PointerEventButton.left) {
                    if(event.kind == PointerEventKind.down) {
                        root.mouse_down_pos.x = self.mouse-x;
                        root.mouse_down_pos.y = self.mouse-y;
                    }
                }
            }
        }
    }
}
```
I know that if the TouchArea is in the outer of the drag_box , it will work very well. However, it is impossible if I want to drag the Window.

There seems to be a solution in https://github.com/slint-ui/slint/pull/2558, but it does not work for me.

**Comment by Vadoola** [comment, marked as answer]
I have a "borderless" draggable window in my project, and it seems to work well. If I recall correctly I was initially using set_position, and got it working, except under Wayland because set_position wasn't supported under Wayland.

I don't know if I have time to try and type up how I got it working, but the repository is public (https://github.com/Vadoola/Tomotroid)

The main files to look at would be

- https://github.com/Vadoola/Tomotroid/blob/main/ui/borderless-window.slint
- https://github.com/Vadoola/Tomotroid/blob/main/ui/appwindow.slint
- The on_move_window section of the main.rs

If you wanted some info on how I got it working (except in Wayland) using set_position, you would have to dig back through my commit history.

The way I'm doing it now, seems to work well, but requires using the private APIs from the  i-slint-backend-winit crate, which may not be what you want.

Hope this helps point you in the right direction.

**Reply by Horbin-Magician** [reply]
Many thanks! I have digged back through your commit history, and found the correct position.

I did a stupid thing that not refresh the `now_pos` in time. In fact, I don't even need to set a 'now pos' myself.
The final solution is:

``` Rust
fn main() {
    let pin_win = PinWindow::new().unwrap();
    
    pin_win.window().set_position( slint::LogicalPosition::new(0., 0.));

    let pin_win_clone = pin_win.as_weak();
    pin_win.on_mouse_move(move |delta_x, delta_y| {
        let pin_win_clone = pin_win_clone.unwrap();
        let logical_pos = pin_win_clone.window().position().to_logical(pin_win_clone.window().scale_factor());
        pin_win_clone.window().set_position(slint::LogicalPosition::new(logical_pos.x + delta_x, logical_pos.y + delta_y));
    });

    pin_win.run().unwrap();
}

slint::slint! {
    export component PinWindow inherits Window {
        no-frame: true;
        always-on-top: true;
        background: red;

        callback mouse_move(length, length);

        TouchArea {
            moved => {
                root.mouse_move(self.mouse-x - self.pressed-x, self.mouse-y - self.pressed-y);
            }
        }
    }
}
```

PS: One thing I find very interesting is that a lot of people like Pomodoro Timer as a practice for UI framework. Your project is very informative to me.

**Reply by Vadoola** [reply]
Glad I could help.
Yeah it's been an useful learning project, although right now it's mostly a bunch of ui files, with little actual functional programming. I've been a bit busy the past few weeks, but I'm hoping I can find some more time to push ahead with it soon.

**Reply by chonwhite** [reply]
I just tried your solution and discovered that using slint::PhysicalPosition instead of slint:: LogicalPosition would yield a smoother dragging experience. Here's my solution in C++, the slint code is unmodified.
```C++
    ui->on_mouse_move([&](int dx, int dy) {
        slint::PhysicalPosition pos = ui->window().position();
        pos.x += dx;
        pos.y += dy;
        ui->window().set_position(pos);
    });
```

**Reply by LucFauvel** [reply]
Hi all, just adding a bit to the conversation here since on windows using the provided answer you lose nice features like window snapping and virtual desktop dragging. The alternative is to rely on winit directly.

```
use i_slint_backend_winit::WinitWindowAccessor;

fn main() {
    let app = App::new().unwrap();
    let app_clone_pin = app.as_weak();
    app.on_start_drag(move || {
        let app_clone = app_clone_pin.unwrap();
        app_clone.window().with_winit_window(|win| {
            let _ = win.drag_window();
        });
    });
}
```

**Comment by Starchaser0** [comment]
i know this has already been answered, but i figured id help out my fellow rust noobs :) 

wayland compatible custom titlebar: #11342
